lib/puppet/functions/reboot/wait.rb

```diff
@@ -0,0 +1,70 @@
+# Waits for nodes to reboot when executed from within a plan
+#
+# This function assumes that the nodes has just been told to reboot and
+# therefore waits for it to disconnect and reconnect again.
+#
+# This has no valid use outside plans!
+Puppet::Functions.create_function(:'reboot::wait') do
+  # @param targets A TargetSpec containing all targets to wait for
+  # @param params Extra parameters defined as a hash, valid keys are:
+  #   disconnect_wait, reconnect_wait and retry_interval. All values should be
+  #   integers and represent seconds.
+  # @example Wait for some very slow nodes to reboot.
+  #   reboot::wait($nodes, { 'disconnect_wait' => 120, 'reconnect_wait' => 600 })
+  dispatch :wait do
+    required_param 'Tuple', :targets
+    optional_param 'Hash', :params
+  end
+
+  def wait(targets, params = { disconnect_wait: 20, reconnect_wait: 120, retry_interval: 1 })
+    threads = []
+
+    targets.each do |target|
+      # We need to thread this so that we can check many nodes at once, it is
+      # possible that this could cause performance issues but only at very
+      # large scales and if you've just triggered 5000 nodes to reboot at once
+      # you have bigger problems...
+      threads << Thread.new do
+        begin
+          # If the target is connected in the beginning, wait for it to disconnect,
+          # then wait for it to come back. Some server may take many minutes to shut
+          # down.
+          wait_until(params[:disconnect_wait], params[:retry_interval]) { !connected?(target) }
```

Comment:
This could be a problem if the system reboots and is `connectable` faster than `retry_interval`.   Not as much of a problem on Windows, but *nix based systems can shutdown/restart very quickly.

That said I'm not sure how to ensure both scenarios apart from querying uptime.

Reply:
> I'd say it's not worth working around

Given that, agreed!